### PR TITLE
Decompile En_Dha

### PR DESCRIPTION
**Pull request by Zelllll**
- Decompile En_Dha
- Changed Link's actor structure
- Changed Global Context structure
- Fixed a bug with GenColliderInit.py

### CODE REVIEW

File: src/overlays/actors/ovl_En_Dha/z_en_dha.h [hunk 1]
```diff
@@ -6,9 +6,28 @@
...
+    /* 0x014C */ SkelAnime skelAnime;
+    /* 0x0190 */ Vec3s limbDrawTable[4];
+    /* 0x01A8 */ Vec3s transitionDrawTable[4];
+    /* 0x01C0 */ u8 unk_1C0;
+    /* 0x01C1 */ char unk_1C1[0x3];
+    /* 0x01C4 */ EnDhaActionFunc actionFunc;
+    /* 0x01C8 */ s16 unk_1C8;
+    /* 0x01CA */ s16 unk_1CA;
+    /* 0x01CC */ u8 unk_1CC;
+    /* 0x01CD */ char unk_1CD[1];
```
Comment (JoshDuMan) [inline src/overlays/actors/ovl_En_Dha/z_en_dha.h:22]: unk_1C1 & unk_1CD are both padding that should still match when removed. They likely are not real fields and are just the gaps.

File: src/overlays/actors/ovl_En_Dha/z_en_dha.c [hunk 2]
```diff
@@ -27,31 +40,367 @@ const ActorInit En_Dha_InitVars = {
...
+                }
+            }
+            func_80035844(&this->unk_1E8, &playerPos, &this->unk_1D0.z, 0);
+            this->unk_1D6.x = ((this->unk_1D6.x - this->actor.shape.rot.y) - this->unk_1D0.y);
+            this->unk_1D0.z = (((this->unk_1D0.z - this->actor.shape.rot.x) - this->unk_1CE) - this->unk_1D0.x);
+        } else {
+            if ((player->stateFlags2 & 0x80) != 0) {
+                if (&this->actor == player->actor.attachedA) {
+                    player->stateFlags2 &= -0x81;
+                    player->actor.attachedA = NULL;
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Dha/z_en_dha.c:182]: youll want to do ~80 instead of -81

**Comment by Zelllll** [comment]
I have fixed every issue mentioned here, hopefully I didn't miss anything. Sorry for all the trouble. :/

File: src/overlays/actors/ovl_En_Dha/z_en_dha.c [hunk 3]
```diff
@@ -15,7 +15,20 @@ void EnDha_Destroy(Actor* thisx, GlobalContext* globalCtx);
...
+void EnDha_SetupAction(EnDha* this, EnDhaActionFunc actionFunc);
+void func_809EC9C8(EnDha* this);
+void func_809ECA50(EnDha* this, GlobalContext* globalCtx);
+void func_809ECF60(EnDha* this);
+void func_809ECF8C(EnDha* this, GlobalContext* globalCtx);
+void EnDha_SetDeadAction(EnDha* this);
+void EnDha_Die(EnDha* this, GlobalContext* globalCtx);
+void EnDha_UpdateHealth(EnDha* this, GlobalContext* globalCtx);
+s32 EnDha_OverrideLimbDraw(GlobalContext* globalCtx, s32 limbIndex, Gfx** dList, Vec3f* pos, Vec3s* rot, Actor* thisx);
+void EnDha_OverridePostDraw(GlobalContext* globalCtx, s32 limbIndex, Gfx** dList, Vec3s* rot, Actor* thisx);
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Dha/z_en_dha.c:27]: Avoid declaring functions like SetupAction or callbacks at the top of the actor unless absolutely necessary. As far as I know, these functions are always defined in an order that doesn't require declaring them at a previous point, so we've been omitting them. Some of the other declarations are probably not necessary either but those are fine imo.
```suggestion
void func_809EC9C8(EnDha* this);
void func_809ECA50(EnDha* this, GlobalContext* globalCtx);
void func_809ECF60(EnDha* this);
void func_809ECF8C(EnDha* this, GlobalContext* globalCtx);
void EnDha_SetDeadAction(EnDha* this);
void EnDha_Die(EnDha* this, GlobalContext* globalCtx);
void EnDha_UpdateHealth(EnDha* this, GlobalContext* globalCtx);
```

File: src/overlays/actors/ovl_En_Dha/z_en_dha.c [hunk 4]
```diff
@@ -27,31 +40,364 @@ const ActorInit En_Dha_InitVars = {
...
+                    this->unk_1CC++;
+                    if (this->actor.attachedA != NULL) {
+                        this->actor.attachedA->params = 1;
+                    }
+                    Audio_PlayActorSound2(&this->actor, NA_SE_EN_DEADHAND_GRIP);
+                }
+            } else {
+                this->unk_1CA += 0x1194;
+                this->unk_1D0.y = (Math_Sins(this->unk_1CA) * 1820.0f);
+                if ((player->stateFlags2 & 0x80) == 0) {
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Dha/z_en_dha.c:163]:
```suggestion
                if (player->stateFlags2 & 0x80) {
```

File: src/overlays/actors/ovl_En_Dha/z_en_dha.c [hunk 5]
```diff
@@ -27,31 +40,364 @@ const ActorInit En_Dha_InitVars = {
...
+                }
+                if (this->unk_1CA < -0x6E6B) {
+                    Audio_PlayActorSound2(&this->actor, NA_SE_EN_DEADHAND_GRIP);
+                }
+            }
+            func_80035844(&this->unk_1E8, &playerPos, &this->unk_1D0.z, 0);
+            this->unk_1D6.x = ((this->unk_1D6.x - this->actor.shape.rot.y) - this->unk_1D0.y);
+            this->unk_1D0.z = (((this->unk_1D0.z - this->actor.shape.rot.x) - this->unk_1CE) - this->unk_1D0.x);
+        } else {
+            if ((player->stateFlags2 & 0x80) != 0) {
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Dha/z_en_dha.c:176]:
```suggestion
            if (player->stateFlags2 & 0x80) {
```

File: src/overlays/actors/ovl_En_Dha/z_en_dha.c [hunk 6]
```diff
@@ -27,31 +40,364 @@ const ActorInit En_Dha_InitVars = {
...
+        func_80035844(&this->unk_1F4, &this->unk_1DC, &unkVar3, 0);
+        Matrix_Translate(this->unk_1DC.x, this->unk_1DC.y, this->unk_1DC.z, 0);
+        Matrix_RotateRPY(unkVar3, unkVar2, 0, 1);
+        Matrix_MultVec3f(&D_809ED758, &this->unk_1F4);
+        Matrix_Translate(this->actor.posRot.pos.x, this->actor.posRot.pos.y, this->actor.posRot.pos.z, 0);
+        func_80035844(&this->actor.posRot, &this->unk_1F4, &unkVar3, 0);
+        Matrix_RotateRPY(unkVar3, unkVar2, 0, 1);
+        Matrix_MultVec3f(&D_809ED74C, &this->unk_1F4);
+        this->unk_1CE = Math_Vec3f_Pitch(&this->actor.posRot, &this->unk_1F4);
+        result = ((Math_Vec3f_Yaw(&this->actor.posRot, &this->unk_1F4) - this->actor.shape.rot.y) << 0x10) >> 0x10;
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Dha/z_en_dha.c:201]: The ` << 0x10) >> 0x10` there is due to a s16 cast, the shifts shouldn't appear in the C code.

File: src/overlays/actors/ovl_En_Dha/z_en_dha.c [hunk 7]
```diff
@@ -27,31 +40,364 @@ const ActorInit En_Dha_InitVars = {
...
+        }
+        this->unk_1D0.x = (Math_Vec3f_Pitch(&this->unk_1F4, &this->unk_1DC) - this->unk_1CE);
+        if (this->unk_1D0.x < 0) {
+            this->unk_1CE += (this->unk_1D0.x * 2);
+            this->unk_1D0.x *= -2;
+            return;
+        }
+    } else {
+        unkVar = -0x81;
+        if ((player->stateFlags2 & 0x80) != 0) {
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Dha/z_en_dha.c:214]:
```suggestion
        if (player->stateFlags2 & 0x80) {
```

File: src/overlays/actors/ovl_En_Dha/z_en_dha.c [hunk 8]
```diff
@@ -27,31 +40,364 @@ const ActorInit En_Dha_InitVars = {
...
+            this->unk_1CE = -0x8000 - this->unk_1CE;
+        }
+        this->unk_1D0.x = (Math_Vec3f_Pitch(&this->unk_1F4, &this->unk_1DC) - this->unk_1CE);
+        if (this->unk_1D0.x < 0) {
+            this->unk_1CE += (this->unk_1D0.x * 2);
+            this->unk_1D0.x *= -2;
+            return;
+        }
+    } else {
+        unkVar = -0x81;
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Dha/z_en_dha.c:213]:
```suggestion
        unkVar = ~0x80;
```

File: src/overlays/actors/ovl_En_Dha/z_en_dha.c [hunk 9]
```diff
@@ -27,31 +40,364 @@ const ActorInit En_Dha_InitVars = {
...
+
+void func_809ECF60(EnDha* this) {
+    this->unk_1C8 = 15;
+    EnDha_SetupAction(this, func_809ECF8C);
+}
+
+void func_809ECF8C(EnDha* this, GlobalContext* globalCtx) {
+    Player* player = PLAYER;
+
+    if ((player->stateFlags2 & 0x80) != 0) {
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Dha/z_en_dha.c:236]:
```suggestion
    if (player->stateFlags2 & 0x80) {
```

File: src/overlays/actors/ovl_En_Dha/z_en_dha.c [hunk 10]
```diff
@@ -27,31 +40,364 @@ const ActorInit En_Dha_InitVars = {
...
+    }
+    EnDha_SetupAction(this, EnDha_Die);
+}
+
+void EnDha_Die(EnDha* this, GlobalContext* globalCtx) {
+    s16 result;
+    Vec3f vector;
+    Player* player = PLAYER;
+
+    if ((player->stateFlags2 & 0x80) != 0) {
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Dha/z_en_dha.c:273]:
```suggestion
    if (player->stateFlags2 & 0x80) {
```

File: src/overlays/actors/ovl_En_Dha/z_en_dha.c [hunk 11]
```diff
@@ -27,31 +40,364 @@ const ActorInit En_Dha_InitVars = {
...
+        if (&this->actor == player->actor.attachedA) {
+            player->stateFlags2 = player->stateFlags2 & -0x81;
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Dha/z_en_dha.c:275]:
```suggestion
            player->stateFlags2 &= ~0x80;
```

File: src/overlays/actors/ovl_En_Dha/z_en_dha.c [hunk 12]
```diff
@@ -27,31 +40,364 @@ const ActorInit En_Dha_InitVars = {
...
+            func_80033480(globalCtx, &vector, 7.0f, 1, 0x5A, 0x14, 1);
+            if (0.0f == this->actor.shape.unk_08) {
+                func_809EC9C8(this);
+            }
+        }
+    }
+}
+
+void EnDha_UpdateHealth(EnDha* this, GlobalContext* globalCtx) {
+    if (!((this->unk_1C0 >= 8) || ((this->collider.base.acFlags & 2) == 0))) {
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Dha/z_en_dha.c:310]:
```suggestion
    if (!((this->unk_1C0 >= 8) || !(this->collider.base.acFlags & 2))) {
```

File: src/overlays/actors/ovl_En_Dha/z_en_dha.c [hunk 13]
```diff
@@ -27,31 +40,364 @@ const ActorInit En_Dha_InitVars = {
...
+        if (this->actor.attachedA->params == 0xA) {
+            EnDha_SetDeadAction(this);
+        }
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Dha/EnDha_Destroy.s")
+void EnDha_Update(Actor* thisx, GlobalContext* globalCtx) {
+    EnDha* this = THIS;
+    s32* colChkCtxTemp;
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Dha/z_en_dha.c:336]:
```suggestion
    CollisionCheckContext* colChkCtx;
```

File: src/overlays/actors/ovl_En_Dha/z_en_dha.c [hunk 14]
```diff
@@ -27,31 +40,364 @@ const ActorInit En_Dha_InitVars = {
...
+            if (this->actor.attachedA != 0) {
+                if (this->actor.attachedA->params == 0xA) {
+                    Actor_Kill(&this->actor);
+                    return;
+                }
+            }
+        } else {
+            this->actor.shape.unk_08 += 500.0f;
+            func_80033480(globalCtx, &vector, 7.0f, 1, 0x5A, 0x14, 1);
+            if (0.0f == this->actor.shape.unk_08) {
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Dha/z_en_dha.c:302]:
```suggestion
            if (this->actor.shape.unk_08 == 0.0f) {
```

File: src/overlays/actors/ovl_En_Dha/z_en_dha.c [hunk 15]
```diff
@@ -27,31 +40,364 @@ const ActorInit En_Dha_InitVars = {
...
+void EnDha_Update(Actor* thisx, GlobalContext* globalCtx) {
+    EnDha* this = THIS;
+    s32* colChkCtxTemp;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Dha/func_809EC9C8.s")
+    colChkCtxTemp = &globalCtx->colChkCtx;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Dha/func_809ECA50.s")
+    if (this->actor.attachedA == NULL) {
+        this->actor.attachedA = Actor_FindNearby(globalCtx, &this->actor, ACTOR_EN_DH, 5, 10000.0f);
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Dha/z_en_dha.c:341]:
```suggestion
        this->actor.attachedA = Actor_FindNearby(globalCtx, &this->actor, ACTOR_EN_DH, ACTORTYPE_ENEMY, 10000.0f);
```

**Comment by Zelllll** [comment]
I have fixed all of the issues mentioned, thanks for the feedback.

File: src/overlays/actors/ovl_En_Dha/z_en_dha.c [hunk 16]
```diff
@@ -27,31 +40,367 @@ const ActorInit En_Dha_InitVars = {
...
+    s32 unkVar;
+    Vec3f playerPos;
+    char pad2[4];
+    s16 unkVar2;
+    s16 unkVar3;
+    s32 result;
+    s32 resultAbs;
+    char pad3[4];
+
+    playerPos = player->actor.posRot.pos;
```
[the remark this answers is not in the document]
Comment (fig02) [reply src/overlays/actors/ovl_En_Dha/z_en_dha.c]: this was marked as resolved but the pads are still chars, i would change those to s32s

File: src/overlays/actors/ovl_En_Dha/z_en_dha.c [hunk 17]
```diff
@@ -27,31 +40,367 @@ const ActorInit En_Dha_InitVars = {
...
+    Math_SmoothScaleMaxMinS(&this->unk_1D0.y, 0, 1, 600, 0);
+    Math_SmoothScaleMaxMinS(&this->unk_1CE, -0x4000, 1, 2000, 0);
+    SkelAnime_FrameUpdateMatrix(&this->skelAnime);
+    this->unk_1C8--;
+    if (this->unk_1C8 == 0) {
+        EnDha_SetAnimation(this);
+    }
+}
+
+void EnDha_SetDeadAction(EnDha* this) {
```
[the remark this answers is not in the document]
Comment (fig02) [reply src/overlays/actors/ovl_En_Dha/z_en_dha.c:250]: this was resolved as well but not changed

**Comment by fig02** [comment]
On top of a few unresolved issues there is also a conflict in undefined syms. These are really easy to resolve, just move it down to some blank lines below

**Pull request merged**